sdks/python/apache_beam/coders/coders.py

```diff
@@ -97,6 +97,16 @@ def is_deterministic(self):
     """
     return False
 
+  def as_deterministic_coder(self, step_label, error_message=None):
```

Comment:
Should the runner guarantee that only one of these will be used ? For example, can a runner encode from non-deterministic one and decode from the corresponding deterministic one. Probably we should tighten the spec more here.

Reply:
This is in the code where coders are being chosen, so there's no separate encode/decode side here, and arbitrary substitution can be done.